### PR TITLE
CRIMAP-615 restructure evidence upload errors

### PR DESCRIPTION
**Pull request by hiboabd**
## Description of change

Restructures the evidence upload error handling to be in line with design patterns. Due to complexities in making the required changes to match designs across js and non-js browsers, the error handling view is amended when javascript is enabled on the browser. When Javascript is not enabled, the error handling will match what was initially implemented (videos below). 

The list of JS changes are: 
- Only list out files in the uploaded files table if they have been successfully uploaded, listing the file names and errors in the error alert component at the top of the page and above the upload component
- Error message content changed including adding the file name to the error message

Other requested changes:
- For non JS evidence upload, remove the dashed border styling as this is for drag and drop components and the non js evidence upload doesn't support drag and drop 
- For non JS evidence upload, do not include the file name error message as the error message is displayed below the file name
- Add a visually hidden 'Delete file' heading in the table column for the delete link for screen readers

## Link to relevant ticket
[Dev ticket - CRIMAP-615](https://dsdmoj.atlassian.net/browse/CRIMAP-615)
[Design ticket - CRIMAP-609](https://dsdmoj.atlassian.net/browse/CRIMAP-609)

## Notes for reviewer
Please test locally and let me know if you spot any weird behaviour 

Also one thing to note is if you were to start uploading on a browser with javascript enabled and then go to a browser with javascript enabled you will go from not seeing error files to seeing them (which could be a quite a list of files, if a person keeps attempting to upload!). This is very unlikely imo but worth mentioning. 

### Before changes:

**Current JS upload error handling staging**

https://github.com/ministryofjustice/laa-apply-for-criminal-legal-aid/assets/51047911/0ece8a30-0904-43e1-b11e-5327caaba856

**Current non JS upload error handling staging**

https://github.com/ministryofjustice/laa-apply-for-criminal-legal-aid/assets/51047911/b5415711-c5a6-421a-bd63-d87a884adfaf

### After changes:

**Js/dropzone upload error handling** 

https://github.com/ministryofjustice/laa-apply-for-criminal-legal-aid/assets/51047911/b217dc0d-bfcf-4c0c-8213-1c106d5ad2aa

**Non js upload error handling**

https://github.com/ministryofjustice/laa-apply-for-criminal-legal-aid/assets/51047911/dff88b1d-0619-4e67-9fde-e11cb77b709c

### CODE REVIEW

File: app/views/steps/evidence/upload/_uploaded_files.html.erb [hunk 1]
```diff
@@ -0,0 +1,16 @@
+<div id="upload-table" class="<%= documents.present? ? '' : 'app-evidence-upload-hidden' %>">
+  <h2 class="govuk-heading-m"> <%= t('steps.evidence.upload.edit.uploaded_files_heading') %></h2>
+  <p class="govuk-body"><%= t('steps.evidence.upload.edit.uploaded_files_info') %></p>
+  <table class="govuk-table">
+    <thead class="govuk-table__head">
+    <tr class="govuk-table__row">
+      <th scope="col" class="govuk-table__header"><%= t('steps.evidence.upload.edit.file_name') %></th>
+      <th scope="col" class="govuk-table__header"><span class="govuk-visually-hidden">Delete file</span></th>
```
Comment (hiboabd) [inline app/views/steps/evidence/upload/_uploaded_files.html.erb:8]: Added a visually hidden 'Delete file' heading for screen readers

File: app/javascript/local/dropzone-cfg.js [hunk 2]
```diff
@@ -2,6 +2,29 @@
 
 import Dropzone from "dropzone"
```
Comment (arthurashman) [inline app/javascript/local/dropzone-cfg.js:4]: I have some concerns that we're now replicating these constants here and in the file_upload_validator. Is there a way that we can have them only in one place?

File: app/javascript/local/dropzone-cfg.js [hunk 3]
```diff
@@ -12,8 +35,23 @@ DropzoneCfg.prototype.init = function () {
...
+
+  this.$chooseFilesButton.addEventListener('click', (e) => {
+    e.preventDefault() // prevent submitting form by default
+  })
+
+  // Display of files table (uploaded files only) is conditional on whether JS is enabled
+  let uploadTableJS = document.getElementById('upload-table-dropzone')
+  uploadTableJS.classList.remove("app-evidence-upload-hidden")
+  let uploadTableNonJS = document.getElementById('upload-table-fallback')
+  uploadTableNonJS.classList.add("app-evidence-upload-hidden")
```
Comment (arthurashman) [inline app/javascript/local/dropzone-cfg.js:54]: If js is not enabled, would this still be able to run?

Comment (hiboabd) [reply app/javascript/local/dropzone-cfg.js:54]: This only needs to run in JS as the default is to make the non js component visible and hide the dropzone component

File: app/javascript/local/dropzone-cfg.js (hunk 2 above)
Comment (hiboabd) [reply app/javascript/local/dropzone-cfg.js:4]: Yeah it's not ideal, i'll explore

File: app/javascript/local/dropzone-cfg.js (hunk 3 above)
Comment (arthurashman) [reply app/javascript/local/dropzone-cfg.js:54]: Oh, I see - I'd misread it. There are 2 tables that you are toggling between

File: app/assets/stylesheets/local/custom.scss [hunk 4]
```diff
@@ -143,3 +143,8 @@ $app-blue-button-colour: govuk-colour("blue");
     }
   }
 }
+
+// To conditionally hide components to support JS/Non JS evidence upload functionality
+.app-evidence-upload-hidden {
+  display: none !important;
+}
```
Comment (mseedat-moj) [inline app/assets/stylesheets/local/custom.scss:150]: Ooooo `!important` is quite drastic - is that absolutely necessary?

File: app/javascript/local/dropzone-cfg.js [hunk 5]
```diff
@@ -113,4 +170,48 @@ function amendErrorLink (file, response) {
...
+  errorSummary.classList.add('app-evidence-upload-hidden')
+
+  document.querySelector('.govuk-form-group').classList.remove('govuk-form-group--error')
+  document.querySelector('.govuk-error-message').classList.add('app-evidence-upload-hidden')
+  document.querySelector('.govuk-error-message').innerHTML = ""
+}
+
+function displayErrorMessage(errorSummary, errorMsg) {
+  // Display error to top of page
+  let pageDiv = document.querySelector('.govuk-grid-column-two-thirds')
```
Comment (mseedat-moj) [inline app/javascript/local/dropzone-cfg.js:189]: Uncomfortable using existing styling classnames as the selector - prefer using html `id` attributes or a dedicated `js-notification-banner`  to make the element selection more specific.

**Review by mseedat-moj — COMMENTED** [review]
Some minor comments from me

File: app/assets/stylesheets/local/custom.scss (hunk 4 above)
Comment (hiboabd) [reply app/assets/stylesheets/local/custom.scss:150]: Actually no 😄 have removed

File: app/javascript/local/dropzone-cfg.js (hunk 2 above)
Comment (hiboabd) [reply app/javascript/local/dropzone-cfg.js:4]: Note: We considered accessing the rails constants by setting a data attribute in the html and then retrieving and using that in the JS. More information [here](https://github.com/rails/webpacker/issues/1684). However, this approach is  recommended against and so we have decided to leave as is

**Review by arthurashman — APPROVED** [review]

**Review by willmcb — APPROVED** [review]
👍 brilliant well done @hiboabd :)

**Pull request merged**